File: app/src/main/java/at/shockbytes/dante/ui/fragment/DownloadBookFragment.java
package at.shockbytes.dante.ui.fragment;


import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.OvershootInterpolator;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

import javax.inject.Inject;

import at.shockbytes.dante.R;
import at.shockbytes.dante.adapter.BookAdapter;
import at.shockbytes.dante.core.DanteApplication;
import at.shockbytes.dante.ui.activity.DownloadActivity;
import at.shockbytes.dante.util.books.Book;
import at.shockbytes.dante.util.books.BookManager;
import at.shockbytes.dante.util.books.BookSuggestion;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import rx.functions.Action1;


public class DownloadBookFragment extends Fragment
        implements Callback, Palette.PaletteAsyncListener, BookAdapter.OnItemClickListener {

    public interface OnBookDownloadedListener {

        void onBookDownloaded(Book book);

        void onCancelDownload();

        void onErrorDownload(String reason);

        void onCloseOnError();

    }

    private static final String ARG_BARCODE = "arg_barcode";

    public static DownloadBookFragment newInstance(String barcode) {
        DownloadBookFragment fragment = new DownloadBookFragment();
        Bundle args = new Bundle();
        args.putString(ARG_BARCODE, barcode);
        fragment.setArguments(args);
        return fragment;
    }

    @Bind({ R.id.fragment_download_book_btn_upcoming,
            R.id.fragment_download_book_btn_current,
            R.id.fragment_download_book_btn_done,
            R.id.fragment_download_book_btn_not_my_book})
    List<View> animViews;

    @Bind(R.id.fragment_download_book_imgview_cover)
    protected ImageView imgViewCover;

    @Bind(R.id.fragment_download_book_txt_title)
    protected TextView txtTitle;

    @Bind(R.id.fragment_download_book_rv_other_suggestions)
    protected RecyclerView rvOtherSuggestions;

    @Bind(R.id.fragment_download_book_txt_other_suggestions)
    protected TextView txtOtherSuggestions;

    @Bind(R.id.fragment_download_book_btn_not_my_book)
    protected Button btnNotMyBook;

    @Bind(R.id.fragment_download_book_progressbar)
    protected ProgressBar progressBar;

    @Bind(R.id.fragment_download_book_main_view)
    protected View mainView;

    @Bind(R.id.fragment_download_book_error_view)
    protected View errorView;

    @Bind(R.id.fragment_download_book_txt_error_cause)
    protected TextView txtErrorCause;

    @Inject
    protected BookManager bookManager;

    private BookAdapter bookAdapter;

    private String query;

    private OnBookDownloadedListener listener;

    private Book selectedBook;

    private boolean isOtherSuggestionsShowing;

    public DownloadBookFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((DanteApplication) getActivity().getApplication()).getAppComponent().inject(this);
        if (getArguments() != null) {
            query = getArguments().getString(ARG_BARCODE);
        }
        isOtherSuggestionsShowing = false;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_download_book, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        downloadBook();
    }

    @Override
    public void onSuccess() {
        Bitmap bm = ((BitmapDrawable) imgViewCover.getDrawable()).getBitmap();
        if (bm != null) {
            Palette.from(bm).generate(this);
        }
    }

    @Override
    public void onError() {

    }

    @Override
    public void onGenerated(Palette palette) {

        // Check for NullPointer
        if (palette.getLightMutedSwatch() == null || palette.getDarkMutedSwatch() == null) {
            return;
        }

        int actionBarColor = palette.getLightMutedSwatch().getRgb();
        int actionBarTextColor = palette.getLightMutedSwatch().getTitleTextColor();
        int statusBarColor = palette.getDarkMutedSwatch().getRgb();

        ((DownloadActivity) getActivity()).colorSystemBars(actionBarColor, actionBarTextColor,
                statusBarColor, selectedBook.getTitle());
    }

    @Override
    public void onItemClick(Book t, View v) {
        int index = bookAdapter.getLocation(t);
        bookAdapter.deleteEntity(t);
        bookAdapter.addEntity(index, selectedBook);

        selectedBook = t;
        setTitleAndIcon(selectedBook);
        rvOtherSuggestions.scrollToPosition(0);
    }


    @OnClick(R.id.fragment_download_book_btn_error_close)
    protected void onClickCloseError() {
        listener.onCloseOnError();
    }

    @OnClick(R.id.fragment_download_book_btn_not_my_book)
    protected void onClickNotMyBook() {

        if (!isOtherSuggestionsShowing) {

            txtTitle.animate().translationY(0).setDuration(500)
                    .setInterpolator(new AccelerateDecelerateInterpolator())
                    .withEndAction(new Runnable() {
                @Override
                public void run() {
                    rvOtherSuggestions.animate().alpha(1).start();
                    txtOtherSuggestions.animate().alpha(1).start();
                }
            }).start();
            imgViewCover.animate().translationY(0).setDuration(500)
                    .setInterpolator(new AccelerateDecelerateInterpolator())
                    .start();

            btnNotMyBook.setText(R.string.download_suggestions_none);

            isOtherSuggestionsShowing = true;
        } else {
            listener.onCancelDownload();
        }

    }

    @OnClick(R.id.fragment_download_book_btn_upcoming)
    protected void onClickUpcoming() {
        finishBookDownload(Book.State.READ_LATER);
    }

    @OnClick(R.id.fragment_download_book_btn_current)
    protected void onClickCurrent() {
        finishBookDownload(Book.State.READING);
    }

    @OnClick(R.id.fragment_download_book_btn_done)
    protected void onClickDone() {
        finishBookDownload(Book.State.READ);
    }

    public void setOnBookDownloadedListener(OnBookDownloadedListener listener) {
        this.listener = listener;
    }

    private void finishBookDownload(Book.State bookState) {

        // Set the state and store it in database
        selectedBook.setState(bookState);
        selectedBook = bookManager.addBook(selectedBook); // Return the object with set ID

        listener.onBookDownloaded(selectedBook);
    }

    private void downloadBook() {

        bookManager.downloadBook(query).subscribe(new Action1<BookSuggestion>() {
            @Override
            public void call(final BookSuggestion suggestion) {

                animateBookViews();
                if (suggestion != null && suggestion.hasSuggestions()) {
                    selectedBook = suggestion.getMainSuggestion();
                    setTitleAndIcon(selectedBook);
                    setupOtherSuggestionsRecyclerView(suggestion.getOtherSuggestions());
                } else {
                    listener.onErrorDownload("no suggestions");
                    showErrorLayout(getString(R.string.download_book_json_error));
                }
            }
        }, new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                throwable.printStackTrace();
                showErrorLayout(throwable);
                listener.onErrorDownload(throwable.getMessage());
            }
        });
    }

    private void animateBookViews() {

        // Hide progressbar smoothly
        progressBar.animate().alpha(0).scaleY(0.5f).scaleX(0.5f).setDuration(500).start();

        ButterKnife.apply(animViews, new ButterKnife.Action<View>() {
            @Override
            public void apply(View view, int index) {
                view.animate().scaleY(1f).scaleX(1f).alpha(1)
                        .setInterpolator(new OvershootInterpolator(4f))
                        .setStartDelay(index * 150).setDuration(250).start();
            }
        });

    }

    private void setTitleAndIcon(Book mainBook) {

        txtTitle.setText(mainBook.getTitle());

        if (mainBook.getThumbnailAddress() != null && !mainBook.getThumbnailAddress().isEmpty()) {
            Picasso.with(getContext()).load(mainBook.getThumbnailAddress())
                    .placeholder(R.drawable.ic_placeholder).into(imgViewCover, this);
        } else {
            imgViewCover.setImageResource(R.drawable.ic_placeholder);
        }
    }

    private void setupOtherSuggestionsRecyclerView(List<Book> books) {

        bookAdapter = new BookAdapter(getContext(), books,
                Book.State.READ, null, false);
        rvOtherSuggestions.setLayoutManager(
                new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        bookAdapter.setOnItemClickListener(this);
        rvOtherSuggestions.setAdapter(bookAdapter);
    }

    private void showErrorLayout(Throwable error) {

        String cause = getString(R.string.download_code_error);
        showErrorLayout(cause);
    }

    private void showErrorLayout(String cause) {
        mainView.setVisibility(View.GONE);
        errorView.setVisibility(View.VISIBLE);
        errorView.animate().alpha(1).start();
        txtErrorCause.setText(cause);
    }


}

File: app/src/main/java/at/shockbytes/dante/ui/fragment/DownloadBookFragment.kt
package at.shockbytes.dante.ui.fragment


import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v7.graphics.Palette
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.view.animation.AnticipateInterpolator
import android.view.animation.OvershootInterpolator
import android.widget.TextView
import at.shockbytes.dante.R
import at.shockbytes.dante.book.BookEntity
import at.shockbytes.dante.book.BookState
import at.shockbytes.dante.dagger.AppComponent
import at.shockbytes.dante.data.BookEntityDao
import at.shockbytes.dante.network.BookDownloader
import at.shockbytes.dante.ui.adapter.BookAdapter
import at.shockbytes.dante.ui.image.ImageLoader
import at.shockbytes.dante.ui.image.ImageLoadingCallback
import at.shockbytes.dante.util.DanteUtils
import at.shockbytes.dante.util.addTo
import at.shockbytes.util.adapter.BaseAdapter
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_download_book.*
import timber.log.Timber
import javax.inject.Inject


class DownloadBookFragment : BaseFragment(), ImageLoadingCallback,
        Palette.PaletteAsyncListener, BaseAdapter.OnItemClickListener<BookEntity> {

    interface OnBookDownloadedListener {

        fun onBookDownloaded(book: BookEntity)

        fun onCancelDownload()

        fun onErrorDownload(reason: String, isAttached: Boolean)

        fun onCloseOnError()

        fun colorSystemBars(actionBarColor: Int?, actionBarTextColor: Int?,
                            statusBarColor: Int?, title: String?)

    }

    private val drawableResList: List<Pair<Int, TextView>> by lazy {
        listOf(Pair(R.drawable.ic_pick_upcoming, btnDownloadFragmentUpcoming),
                Pair(R.drawable.ic_pick_current, btnDownloadFragmentCurrent),
                Pair(R.drawable.ic_pick_done, btnDownloadFragmentDone))
    }

    private val animViews: List<View> by lazy {
        listOf(btnDownloadFragmentUpcoming,
                btnDownloadFragmentCurrent,
                btnDownloadFragmentDone,
                btnDownloadFragmentNotMyBook)
    }

    @Inject
    protected lateinit var bookDownloader: BookDownloader

    @Inject
    protected lateinit var bookDao: BookEntityDao

    @Inject
    protected lateinit var imageLoader: ImageLoader

    private var bookAdapter: BookAdapter? = null
    private var listener: OnBookDownloadedListener? = null
    private var selectedBook: BookEntity? = null
    private var isOtherSuggestionsShowing: Boolean = false
    private var query: String? = null

    override val layoutId = R.layout.fragment_download_book

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        query = arguments?.getString(argBarcode)
        isOtherSuggestionsShowing = false
    }

    override fun injectToGraph(appComponent: AppComponent) {
        appComponent.inject(this)
    }

    override fun setupViews() {
        loadIcons()
        downloadBook()

        btnDownloadFragmentErrorClose.setOnClickListener {
            listener?.onCloseOnError()
        }
        btnDownloadFragmentNotMyBook.setOnClickListener {
            showOtherSuggestions()
        }
        btnDownloadFragmentUpcoming.setOnClickListener {
            finishBookDownload(BookState.READ_LATER)
        }
        btnDownloadFragmentCurrent.setOnClickListener {
            finishBookDownload(BookState.READING)
        }
        btnDownloadFragmentDone.setOnClickListener {
            finishBookDownload(BookState.READ)
        }

    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        listener = context as? OnBookDownloadedListener
    }


    override fun onImageLoadingFailed(e: Exception?) {
        Timber.e(e)
    }

    override fun onImageResourceReady(resource: Drawable?) {
        (resource as? BitmapDrawable)?.bitmap?.let {
            Palette.from(it).generate(this)
        }
    }


    override fun onGenerated(palette: Palette?) {

        val actionBarColor = palette?.lightMutedSwatch?.rgb
        val actionBarTextColor = palette?.lightMutedSwatch?.titleTextColor
        val statusBarColor = palette?.darkMutedSwatch?.rgb

        listener?.colorSystemBars(actionBarColor, actionBarTextColor,
                statusBarColor, selectedBook?.title)
    }

    override fun onItemClick(t: BookEntity, v: View) {
        val index = bookAdapter?.getLocation(t) ?: return
        bookAdapter?.deleteEntity(t)
        bookAdapter?.addEntity(index, selectedBook!!)

        selectedBook = t
        setTitleAndIcon(selectedBook)
        recyclerViewDownloadFragmentOtherSuggestions.scrollToPosition(0)
    }

    // --------------------------------------------------------------------

    private fun showOtherSuggestions() {

        if (!isOtherSuggestionsShowing) {

            txtDownloadFragmentTitle.animate().translationY(0f).setDuration(300)
                    .setInterpolator(AnticipateInterpolator())
                    .withEndAction {
                        recyclerViewDownloadFragmentOtherSuggestions.animate().alpha(1f).start()
                        txtDownloadFragmentOtherSuggestions.animate().alpha(1f).start()
                    }.start()
            imgViewDownloadFragmentCover.animate().translationY(0f).setDuration(300)
                    .setInterpolator(AnticipateInterpolator())
                    .start()

            btnDownloadFragmentNotMyBook.setText(R.string.download_suggestions_none)

            isOtherSuggestionsShowing = true
        } else {
            listener?.onCancelDownload()
        }
    }

    private fun finishBookDownload(bookState: BookState) {
        selectedBook?.let { book ->
            // Set the state and store it in database
            book.updateState(bookState)
            bookDao.create(book)
            listener?.onBookDownloaded(book)
        }
    }

    private fun downloadBook() {
        query?.let { q ->
            bookDownloader.downloadBook(q).subscribe({ suggestion ->
                animateBookViews()
                if (suggestion != null && suggestion.hasSuggestions) {
                    selectedBook = suggestion.mainSuggestion
                    setTitleAndIcon(selectedBook)
                    setupOtherSuggestionsRecyclerView(suggestion.otherSuggestions)
                    activity?.actionBar?.title = selectedBook?.title
                } else {
                    listener?.onErrorDownload("no suggestions", isAdded)
                    showErrorLayout(getString(R.string.download_book_json_error))
                }
            }) { throwable: Throwable? ->
                throwable?.printStackTrace()
                showErrorLayout(throwable)
                listener?.onErrorDownload(throwable?.localizedMessage
                        ?: "Error message not available", isAdded)
            }
        }
    }

    private fun loadIcons() {
        Single.fromCallable {
            drawableResList.mapNotNull { (drawableRes, view) ->
                context?.let { ctx ->
                    val drawable = DanteUtils.vector2Drawable(ctx, drawableRes)
                    Pair(drawable, view)
                }
            }
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe { list ->
            list.forEach { (drawable, view) ->
                view.setCompoundDrawablesWithIntrinsicBounds(null, drawable, null, null)
            }
        }.addTo(compositeDisposable)
    }

    private fun animateBookViews() {
        // Hide progressbar smoothly
        progressBarDownloadFragment.animate()
                .alpha(0f)
                .scaleY(0.5f)
                .scaleX(0.5f)
                .setDuration(500)
                .start()
        DanteUtils.listPopAnimation(animViews, 250, interpolator = OvershootInterpolator(4f))
    }

    private fun setTitleAndIcon(mainBook: BookEntity?) {

        txtDownloadFragmentTitle.text = mainBook?.title

        val address = mainBook?.thumbnailAddress
        if (!address.isNullOrEmpty()) {
            context?.let { ctx ->
                imageLoader.loadImage(ctx, address!!, imgViewDownloadFragmentCover,
                        callback = this, callbackHandleValues = Pair(false, true))
            }
        } else {
            imgViewDownloadFragmentCover.setImageResource(R.drawable.ic_placeholder)
        }
    }

    private fun setupOtherSuggestionsRecyclerView(books: List<BookEntity>) {

        bookAdapter = BookAdapter(context!!, books, BookState.READ, imageLoader,null, false)
        recyclerViewDownloadFragmentOtherSuggestions.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        bookAdapter?.onItemClickListener = this
        recyclerViewDownloadFragmentOtherSuggestions.adapter = bookAdapter
    }

    private fun showErrorLayout(error: Throwable?) {
        error?.let {
            Timber.e(it)
        }

        if (isAdded) {
            val cause = getString(R.string.download_code_error)
            showErrorLayout(cause)
        }
    }

    private fun showErrorLayout(cause: String) {
        if (isAdded) {
            layoutDownloadFragmentMain.visibility = View.GONE
            layoutDownloadFragmentError.visibility = View.VISIBLE
            layoutDownloadFragmentError.animate().alpha(1f).start()
            txtDownloadFragmentErrorCause.text = cause
        } else {
            showToast(cause, true)
            // Log this message, because this should not happen
            Timber.e(IllegalArgumentException("Cannot show error layout, because DownloadBookFragment is not attached to Activity"))
        }

    }

    // --------------------------------------------------------------------

    companion object {

        private const val argBarcode = "arg_barcode"

        fun newInstance(barcode: String?): DownloadBookFragment {
            val fragment = DownloadBookFragment()
            val args = Bundle()
            args.putString(argBarcode, barcode)
            fragment.arguments = args
            return fragment
        }
    }


}
